omnibus/files/private-chef-cookbooks/private-chef/resources/pg_upgrade.rb

```diff
@@ -237,7 +235,7 @@ def update_to_latest_version
 
         # The new data directory may have just been created; that's why
         # this needs to be evaluated lazily
-        new_version = version_from_data_dir(new_data_dir)
+        new_version = node['private_chef']['postgresql']['version']
```

Comment:
Any reason the old code does not work?

Reply:
See commit message below:

```
        * fix new_version

        pg_upgrade.rb was calculating incorrect paths for the pg_upgrade
    command - specifically
        the path to execute the command itself, and the path for the
    --new-bindir flag.  this was
        because postgresql changed version numbering with release !),
    affecting
        how we calculate these paths.

        context:

        https://www.postgresql.org/support/versioning/

        Version Numbering

        Starting with PostgreSQL !), a major version is indicated by
    increasing the first part of the version, e.g. !) to !!. Before
    PostgreSQL !), a major versio
    n was indicated by increasing either the first or second part of the
    version number, e.g. (.% to (.^.

        Minor releases are numbered by increasing the last part of the
    version number. Beginning with PostgreSQL !), this is the second part of
    the version number
    , e.g. !).) to !).!; for older versions this is the third part of the
    version number, e.g. (.%.# to (.%.$.

        https://pgpedia.info/p/PG_VERSION.html

        In PostgreSQL (.^ and earlier, the value shows the version expressed
    in series.major_release format
        In PostgreSQL !) and later, the version is always a single number

        error:

        [#!mError executing action `run` on resource
    'execute[upgrade_postgres_cluster]'

        null_resource.chef_server_upgrade 9remote-exec0:
    ================================================================================

        null_resource.chef_server_upgrade 9remote-exec0:
    Mixlib::ShellOut::ShellCommandFailed

        null_resource.chef_server_upgrade 9remote-exec0:
    ------------------------------------

        null_resource.chef_server_upgrade 9remote-exec0:       Expected
    process to exit with [)], but received '!@&'

        null_resource.chef_server_upgrade 9remote-exec0:       ---- Begin
    output of /opt/opscode/embedded/postgresql/!@/bin/pg_upgrade
    --old-datadir=/var/opt/opsc
    ode/postgresql/(.^/data
    --new-datadir=/var/opt/opscode/postgresql/!@.%/data
    --old-bindir=/opt/opscode/embedded/postgresql/(.^/bin
    --new-bindir=/opt/opscode/em
    bedded/postgresql/!@/bin --old-options=" -c
    config_file=/var/opt/opscode/postgresql/(.^/data/postgresql.conf"
    --new-options=" -c config_file=/var/opt/opscode/
    postgresql/!@.%/data/postgresql.conf" 77 date >
    /var/opt/opscode/postgresql/!@.%/data/upgraded.sentinel ----

        null_resource.chef_server_upgrade 9remote-exec0:       STDOUT:

        null_resource.chef_server_upgrade 9remote-exec0:       STDERR: sh:
    !: /opt/opscode/embedded/postgresql/!@/bin/pg_upgrade: not found

        null_resource.chef_server_upgrade 9remote-exec0:       ---- End
    output of /opt/opscode/embedded/postgresql/!@/bin/pg_upgrade
    --old-datadir=/var/opt/opscod
    e/postgresql/(.^/data
    --new-datadir=/var/opt/opscode/postgresql/!@.%/data
    --old-bindir=/opt/opscode/embedded/postgresql/(.^/bin
    --new-bindir=/opt/opscode/embe
    dded/postgresql/!@/bin --old-options=" -c
    config_file=/var/opt/opscode/postgresql/(.^/data/postgresql.conf"
    --new-options=" -c config_file=/var/opt/opscode/po
    stgresql/!@.%/data/postgresql.conf" 77 date >
    /var/opt/opscode/postgresql/!@.%/data/upgraded.sentinel ----

        null_resource.chef_server_upgrade 9remote-exec0:       Ran
    /opt/opscode/embedded/postgresql/!@/bin/pg_upgrade
    --old-datadir=/var/opt/opscode/postgresql/(.
    ^/data --new-datadir=/var/opt/opscode/postgresql/!@.%/data
    --old-bindir=/opt/opscode/embedded/postgresql/(.^/bin
    --new-bindir=/opt/opscode/embedded/postgresql
    /!@/bin --old-options=" -c
    config_file=/var/opt/opscode/postgresql/(.^/data/postgresql.conf"
    --new-options=" -c config_file=/var/opt/opscode/postgresql/!@.%/d
    ata/postgresql.conf" 77 date >
    /var/opt/opscode/postgresql/!@.%/data/upgraded.sentinel returned !@&

        null_resource.chef_server_upgrade 9remote-exec0:       Resource
    Declaration:

        null_resource.chef_server_upgrade 9remote-exec0:
    ---------------------

        null_resource.chef_server_upgrade 9remote-exec0:       3 In
    /var/opt/opscode/local-mode-cache/cookbooks/private-chef/providers/pg_upgrade.rb

        Signed-off-by: Lincoln Baker <lbaker2chef.io>

    diff --git
    a/omnibus/files/private-chef-cookbooks/private-chef/providers/pg_upgrade.rb
    b/omnibus/files/private-chef-cookbooks/private-chef/providers/pg_upgrad
    e.rb
    index @$%a%d@&^..c!cdca)e! !))^$$
    ---
    a/omnibus/files/private-chef-cookbooks/private-chef/providers/pg_upgrade.rb
    +++
    b/omnibus/files/private-chef-cookbooks/private-chef/providers/pg_upgrade.rb
    22 -@#%,& +@#%,& 22 def update_to_latest_version

           3 The new data directory may have just been created; that's why
           3 this needs to be evaluated lazily
    -      new_version = version_from_data_dir9new_data_dir0
    +      new_version = node['private_chef']['postgresql']['version']

           old_bins = binary_path_for9old_version0
           new_bins = binary_path_for9new_version0
```